vm-virtio/src/queue.rs

```diff
@@ -29,6 +29,37 @@ pub(super) const VIRTQ_DESC_F_WRITE: u16 = 0x2;
 // The Virtio Spec 1.0 defines the alignment of VirtIO descriptor is 16 bytes,
 // which fulfills the explicit constraint of GuestMemoryMmap::read_obj().
 
+/// An iterator over a single descriptor chain.  Not to be confused with AvailIter,
+/// which iterates over the descriptor chain heads in a queue.
+pub struct DescIter<'a> {
+    next: Option<DescriptorChain<'a>>,
+}
+
+impl<'a> DescIter<'a> {
+    /// Returns an iterator that only yields the readable descriptors in the chain.
+    pub fn readable(self) -> impl Iterator<Item = DescriptorChain<'a>> {
+        self.skip_while(DescriptorChain::is_write_only)
+    }
+
+    /// Returns an iterator that only yields the writable descriptors in the chain.
+    pub fn writable(self) -> impl Iterator<Item = DescriptorChain<'a>> {
+        self.take_while(DescriptorChain::is_write_only)
```

Comment:
I think this might be your bug.  `take_while()` will ignore all remaining elements in the iterator after the first time the closure returns false and the virtio spec requires that the driver put all the readable descriptors in the chain before the writable descriptors.

So `DescriptorChain::is_write_only` will return false on the first descriptor in the chain (because always a read-only descriptor) and then ignore everything else in the iterator.  The crosvm code for this function is actually:

```rust
self.skip_while(DescriptorChain::is_read_only)
```

If the `DescriptorChain` struct doesn't have an `is_read_only()` method this could also be something like 

```rust
self.filter(DescriptorChain::is_write_only)
```
 or

```rust
self.skip_while(|d| !d.is_write_only())
```

Reply:
Yes that's exactly what I found yesterday, and the fix is part of the `debug` commit. I didn't realize the behavior of `skip_while` before :(
Anyway, thank you very much for your time looking at this :)